sites/cnfpt/src/backend/base/tests/test_cache.py

```diff
@@ -0,0 +1,135 @@
+    def test_get_fallback_cache(
+        self, redis_cache_mock, fallback_cache_mock, logger_mock
+    ):
+        """
+        Test case when redis_cache raises an exception,
+        logger logs the exception then fallback_cache takes over
+        """
+        client = RedisCacheWithFallback(None, {})
+        client.get("irrelevent")
+
+        redis_cache_mock.assert_called_once()
+        fallback_cache_mock.assert_not_called()
+
+        redis_cache_mock.side_effect = Exception()
+        client.get("irrelevent")
+
+        logger_mock.warning.assert_called_with(
+            "[DEGRADED CACHE MODE] - Switch to fallback cache"
+        )
+        logger_mock.exception.assert_called_once()
+        fallback_cache_mock.assert_called_once()
+
+    @override_settings(
+        CACHES={
+            "default": {
+                "BACKEND": "base.cache.RedisCacheWithFallback",
+                "LOCATION": "mymaster/redis-sentinel:26379,redis-sentinel:26379/0",
+                "OPTIONS": {
+                    "CLIENT_CLASS": "richie.apps.core.cache.SentinelClient",
+                },
+            },
+            "fallback": {
+                "BACKEND": "django.core.cache.backends.dummy.DummyCache",
+            },
+        }
+    )
+    @mock.patch.object(DummyCache, "delete")
+    @mock.patch.object(RedisCacheWithFallback, "_call_redis_cache")
+    def test_invalidate_fallback_cache(self, redis_cache_mock, delete_mock):
+        """
+        Test that fallback_cache is invalidated every 60 seconds
+        when Redis is up
+        """
+        client = RedisCacheWithFallback(None, {})
+        client.get("irrelevant")
+
+        redis_cache_mock.assert_called_once()
+        delete_mock.assert_not_called()
+        redis_cache_mock.reset_mock()
+        delete_mock.reset_mock()
+
+        now = datetime.datetime.now() + datetime.timedelta(seconds=120)
+        with mock.patch("base.utils.datetime") as mock_datetime:
+            mock_datetime.now.return_value = now
+            client.get("irrelevant")
+
+        redis_cache_mock.assert_called_once()
+        delete_mock.assert_called_once()
+        redis_cache_mock.reset_mock()
+        delete_mock.reset_mock()
+
+        now = datetime.datetime.now() + datetime.timedelta(seconds=30)
+        with mock.patch("base.utils.datetime") as mock_datetime:
+            mock_datetime.now.return_value = now
+            client.get("irrelevant")
+
+        redis_cache_mock.assert_called_once()
+        delete_mock.assert_not_called()
+        redis_cache_mock.reset_mock()
+        delete_mock.reset_mock()
+
+        now += datetime.timedelta(seconds=30)
+        with mock.patch("base.utils.datetime") as mock_datetime:
+            mock_datetime.now.return_value = now
+            client.get("irrelevant")
+
+        redis_cache_mock.assert_called_once()
+        delete_mock.assert_not_called()
+        redis_cache_mock.reset_mock()
+        delete_mock.reset_mock()
```

Comment:
I have 2 suggestions for improvement:
- intialize `now = datetime.now()` only once at the beginning of the test and use another variable `read_time = now + timedelta(seconds=30)`. I know the execution time is negligeable but it seems that this last case you're testing should then be >60 seconds and therefore trigger a new delete?
- get a different key each time to make sure that the throttle is independant of what is retrieved in cache.

Reply:
Very good idea, it avoids some big mistakes.